### ISSUE TITLE
The rule for using quotes only when they are really needed

### ISSUE THREAD

**Issue by scherebedov** [adrienverge/yamllint#133]
Is it possible to make a rule to use quotes only when they are really needed?
Like a [this](https://stackoverflow.com/questions/19109912/do-i-need-quotes-for-strings-in-yaml?answertab=votes#tab-top) in answer.
Thank's.

**Comment by adrienverge** [comment]
Hi,

Can you be more explicit on when quotes are "really needed"? The link you provide doesn't describe what you want.

For instance:
```yaml
boolean: yes          # I don't want quotes, it's a boolean
number: 10            # I don't want quotes, it's a number
dict: {}              # I don't want quotes, it's an object
string: abc 10 {}     # I don't need quotes, it's a string
```

Please provide explicit examples of what you would like this rule to do.

Also, for info, since version 1.12.0 the `quoted-strings` rule is available. But I don't think this is exactly what you're looking for: https://yamllint.readthedocs.io/en/latest/rules.html#module-yamllint.rules.quoted_strings

**Comment by adrienverge** [comment]
@dctr that makes sense. I'll review contributions to add a new option.

**Comment by tavin** [comment]
@adrienverge does that mean you'd like to see a PR?
I wonder if it should be handled by `quoted-strings` or by its own rule...

**Comment by adrienverge** [comment]
Yes it means that if someones wants to implement this, I'll review the PR. In my opinion it makes most sense to be handled inside `quoted-strings`.

**Comment by ruipinge** [comment]
Although I created a PR with the proposed solution, I believe there's a better one with more flexibility for this rule. Instead of `allow-redundant` something like `usage` with options like:
- `all`: any string value would need to be quoted using `quote-type` definition (currently the default, my suggestion would be to keep it)
- `whatever`: quotes are not mandatory, but when used must follow `quote-type` definition
- `no-redundant`: disallow unneeded quotes, but when needed must follow `quote-type` definition

Any suggestion on different naming are welcome.

If there's an agreement, I'll be glad to update the PR.

**Comment by ruipinge** [comment]
Since they are related, this PR already includes the #233 fix.

@adrienverge any chance we can have your feedback on this?

**Comment by adrienverge** [comment]
@ruipinge thanks for taking care of that. I agree your proposal gives more flexibility and is more future-proof.

About the naming, what do you think of:
```yaml
quoted-strings:
  quote-type: single | double | any
  required: true | false | only-if-needed
```
?

(I'm not al English speaker, maybe `only-when-needed` is better than `only-if-needed`?)

**Comment by ruipinge** [comment]
@adrienverge I like your naming suggestions. 👍 I believe `only-when-needed` sounds better.

I'll update my PR accordingly.

**Comment by ruipinge** [comment]
@adrienverge updated the pull request accordingly. Waiting for review.

**Comment by adrienverge** [comment]
Implemented by #235 and released in v1.21.0. See https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.quoted_strings.
